middle_end/flambda2/simplify/inlining/function_decl_inlining_decision.ml

```diff
@@ -17,35 +17,40 @@
 [@@@ocaml.warning "+a-30-40-41-42"]
 
 let make_decision ~inlining_arguments:args ~inline ~stub ~cost_metrics:metrics
-    ~is_a_functor : Function_decl_inlining_decision_type.t =
+    ~is_a_functor ~(is_recursive : Recursive.t) :
+    Function_decl_inlining_decision_type.t =
   (* At present, we follow Closure, taking inlining decisions without first
      examining call sites. *)
   match (inline : Inline_attribute.t) with
   | Never_inline -> Never_inline_attribute
   | Always_inline -> Attribute_inline
-  | Default_inline | Unroll _ | Available_inline ->
+  | Default_inline | Unroll _ | Available_inline -> (
     if stub
     then Stub
     else
-      let large_function_size =
-        Inlining_arguments.large_function_size args |> Code_size.of_int
-      in
-      let small_function_size =
-        Inlining_arguments.small_function_size args |> Code_size.of_int
-      in
-      let size = Cost_metrics.size metrics in
-      let is_small = Code_size.( <= ) size small_function_size in
-      let is_large = Code_size.( <= ) large_function_size size in
-      if is_a_functor
-      then Functor { size }
-      else if is_large && not (Inline_attribute.equal inline Available_inline)
-      then Function_body_too_large large_function_size
-      else if is_small
-      then
-        Small_function { size = Cost_metrics.size metrics; small_function_size }
-      else
-        Speculatively_inlinable
-          { size = Cost_metrics.size metrics;
-            small_function_size;
-            large_function_size
-          }
+      match is_recursive with
+      | Recursive -> Never_inline_attribute
```

Comment:
There should be a `Not_inlining_recursive_functions` decision for this, no? There isn't an attribute involved.

Also, this would seem surprising if there's an unroll attribute. If `@inline always` overrides the command line, I would think so would `@unroll`.

Reply:
I'm not sure it's correct for the command-line attributes to override everything.  I think they should stop at the maximum recursion depth and the maximum inlining depth, no?

I wonder if we could make this decision based on `-inline-max-unroll` -- seems ok, @chambart ?

Reply:
(and yes, this should have a new constructor in the decision type)

Reply:
Oh but hold on: `-inline-max-unroll` is an Flambda 1 not an Flambda 2 option.  In which case adding a new flag for this as @chambart has done seems like the right thing to do -- although... maybe this should actually be controlled by the "max rec depth" parameter?  This is currently hard-coded but is changed to a command-line flag in another PR (I think it's my Simplify CRs one).  Perhaps if max-rec-depth = 0 then we should have this behaviour, max-rec-depth = 1 means that a recursive function may have its body substituted at the call site exactly once, max-rec-depth = 2 means that after such substitution there may be another unrolling, etc.?

Reply:
!! I _thought_ I remembered wanting to change something to a command-line option, then I got confused when I saw stuff about inline_max_unroll.

Yes, that would be my preference: have that be an argument and the default be zero.

Reply:
We discussed this on slack and @chambart said this should be two different options, since the rec-depth one is contextual.